<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 1: Standardized Test Analysis

--- 
# Part 1

Part 1 requires knowledge of basic Python.

---

## Problem Statement

The SAT and ACT are standardized tests used for college admissions in the United States. This project aims to explore trends in SAT and ACT scores and participation for the years 2017-2019 and make recommendations to College Board to improve participation rates of SAT.


### Contents:
- [Background](#Background)
- [Data Import & Cleaning](#Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-Data)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

## Background

The SAT and ACT are standardized tests that many colleges and universities in the United States require for their admissions process. This score is used along with other materials such as grade point average (GPA) and essay responses to determine whether or not a potential student will be accepted to the university.

The SAT has two sections of the test: Evidence-Based Reading and Writing and Math ([*source*](https://www.princetonreview.com/college/sat-sections)). The ACT has 4 sections: English, Mathematics, Reading, and Science, with an additional optional writing section ([*source*](https://www.act.org/content/act/en/products-and-services/the-act/scores/understanding-your-scores.html)). They have different score ranges, which you can read more about on their websites or additional outside sources (a quick Google search will help you understand the scores for each test):
* [SAT](https://collegereadiness.collegeboard.org/sat)
* [ACT](https://www.act.org/content/act/en.html)

Standardized tests have long been a controversial topic for students, administrators, and legislators. Since the 1940's, an increasing number of colleges have been using scores from sudents' performances on tests like the SAT and the ACT as a measure for college readiness and aptitude ([*source*](https://www.minotdailynews.com/news/local-news/2017/04/a-brief-history-of-the-sat-and-act/)). Supporters of these tests argue that these scores can be used as an objective measure to determine college admittance. Opponents of these tests claim that these tests are not accurate measures of students potential or ability and serve as an inequitable barrier to entry.

### Choose your Data

There are 10 datasets included in the [`data`](./data/) folder for this project. You are required to pick **at least two** of these to complete your analysis. Feel free to use more than two if you would like, or add other relevant datasets you find online.

* [`act_2017.csv`](./data/act_2017.csv): 2017 ACT Scores by State
* [`act_2018.csv`](./data/act_2018.csv): 2018 ACT Scores by State
* [`act_2019.csv`](./data/act_2019.csv): 2019 ACT Scores by State
* [`act_2019_ca.csv`](./data/act_2019_ca.csv): 2019 ACT Scores in California by School
* [`sat_2017.csv`](./data/sat_2017.csv): 2017 SAT Scores by State
* [`sat_2018.csv`](./data/sat_2018.csv): 2018 SAT Scores by State
* [`sat_2019.csv`](./data/sat_2019.csv): 2019 SAT Scores by State
* [`sat_2019_by_intended_college_major.csv`](./data/sat_2019_by_intended_college_major.csv): 2019 SAT Scores by Intended College Major
* [`sat_2019_ca.csv`](./data/sat_2019_ca.csv): 2019 SAT Scores in California by School
* [`sat_act_by_college.csv`](./data/sat_act_by_college.csv): Ranges of Accepted ACT & SAT Student Scores by Colleges

*The following datasets will be used for this analysis:*
* [`act_2017.csv`](./data/act_2017.csv): 2017 ACT Scores by State
* [`act_2018.csv`](./data/act_2018.csv): 2018 ACT Scores by State
* [`act_2019.csv`](./data/act_2019.csv): 2019 ACT Scores by State
* [`sat_2017.csv`](./data/sat_2017.csv): 2017 SAT Scores by State
* [`sat_2018.csv`](./data/sat_2018.csv): 2018 SAT Scores by State
* [`sat_2019.csv`](./data/sat_2019.csv): 2019 SAT Scores by State

### Outside Research

The new format for the SAT was released in March 2016. Since then, levels of participation in multiple states have changed with varying legislative decisions. At the same time, College Board has also partnered Khan Academy to provide free SAT preparation, which led to an increase in SAT test-takers ([*source*](https://newsroom.collegeboard.org/college-board-announces-surge-students-taking-new-satr-suite-assessments-creating-opportunity)).<br>
In 2019, SAT has took over ACT as the more popular test. More than 2.2 million students in the class of 2019 took the SAT ([*source*](https://newsroom.collegeboard.org/over-22-million-students-class-2019-took-sat-largest-group-ever#:~:text=Over%202.2%20Million%20Students%20in,Group%20Ever%20%7C%20The%20College%20Board)), as compared to 1.8 million students who took the ACT ([*source*](https://www.act.org/content/dam/act/unsecured/documents/National-CCCR-2019.pdf)).<br>
Some companies have been reported to ask job candidates for their SAT scores, which could be used as a differentiator among other candidates ([*source*](https://www.wsj.com/articles/SB10001424052702303636404579395220334268350)). 

### Coding Challenges

1. Manually calculate mean:

    Write a function that takes in values and returns the mean of the values. Create a list of numbers that you test on your function to check to make sure your function works!
    
    *Note*: Do not use any mean methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [1]:
# Code:
def mean_function(num_list):
    mean = sum(num_list) / len(num_list)
    return mean


In [2]:
sample_list = [1,4,6,8]
mean_function(sample_list)

4.75

2. Manually calculate standard deviation:

    The formula for standard deviation is below:

    $$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

    Where $x_i$ represents each value in the dataset, $\mu$ represents the mean of all values in the dataset and $n$ represents the number of values in the dataset.

    Write a function that takes in values and returns the standard deviation of the values using the formula above. Hint: use the function you wrote above to calculate the mean! Use the list of numbers you created above to test on your function.
    
    *Note*: Do not use any standard deviation methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [3]:
def std_dev_function(num_list):
    summation = 0
    for num in num_list:
        summation += (num - mean_function(num_list))**2
    std_dev = (summation / len(num_list))**0.5
    return std_dev


In [4]:
sample_list = [1,4,6,8]
std_dev_function(sample_list)

2.5860201081971503

3. Data cleaning function:
    
    Write a function that takes in a string that is a number and a percent symbol (ex. '50%', '30.5%', etc.) and converts this to a float that is the decimal approximation of the percent. For example, inputting '50%' in your function should return 0.5, '30.5%' should return 0.305, etc. Make sure to test your function to make sure it works!

You will use these functions later on in the project!

In [5]:
def convert_to_float(string):
    new_float = float(string.replace('%', '')) / 100
    return new_float

In [6]:
convert_to_float('30.5%')

0.305

--- 
# Part 2

Part 2 requires knowledge of Pandas, EDA, data cleaning, and data visualization.

---

*All libraries used should be added here*

In [7]:
# Imports:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Data Import and Cleaning

### Data Import & Cleaning

Import the datasets that you selected for this project and go through the following steps at a minimum. You are welcome to do further cleaning as you feel necessary:
1. Display the data: print the first 5 rows of each dataframe to your Jupyter notebook.
2. Check for missing values.
3. Check for any obvious issues with the observations (keep in mind the minimum & maximum possible values for each test/subtest).
4. Fix any errors you identified in steps 2-3.
5. Display the data types of each feature.
6. Fix any incorrect data types found in step 5.
    - Fix any individual values preventing other columns from being the appropriate type.
    - If your dataset has a column of percents (ex. '50%', '30.5%', etc.), use the function you wrote in Part 1 (coding challenges, number 3) to convert this to floats! *Hint*: use `.map()` or `.apply()`.
7. Rename Columns.
    - Column names should be all lowercase.
    - Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`).
    - Column names should be unique and informative.
8. Drop unnecessary rows (if needed).
9. Merge dataframes that can be merged.
10. Perform any additional cleaning that you feel is necessary.
11. Save your cleaned and merged dataframes as csv files.

In [8]:
# importing datasets

act_2017 = pd.read_csv('../data/act_2017.csv') 
act_2018 = pd.read_csv('../data/act_2018.csv')
act_2019 = pd.read_csv('../data/act_2019.csv')
sat_2017 = pd.read_csv('../data/sat_2017.csv')
sat_2018 = pd.read_csv('../data/sat_2018.csv')
sat_2019 = pd.read_csv('../data/sat_2019.csv')

*We will be displaying the first 5 rows of each dataset using .head() function:*

In [9]:
act_2017.head() 

State Participation  English  Math  Reading  Science Composite
0  National           60%     20.3  20.7     21.4     21.0      21.0
1   Alabama          100%     18.9  18.4     19.7     19.4      19.2
2    Alaska           65%     18.7  19.8     20.4     19.9      19.8
3   Arizona           62%     18.6  19.8     20.1     19.8      19.7
4  Arkansas          100%     18.9  19.0     19.7     19.5      19.4

In [10]:
act_2018.head()

State Participation  Composite
0     Alabama          100%       19.1
1      Alaska           33%       20.8
2     Arizona           66%       19.2
3    Arkansas          100%       19.4
4  California           27%       22.7

In [11]:
act_2019.head()

State Participation  Composite
0     Alabama          100%       18.9
1      Alaska           38%       20.1
2     Arizona           73%       19.0
3    Arkansas          100%       19.3
4  California           23%       22.6

*From the datasets above, the English, Math, Reading and Science scores are missing from ACT 2018 and 2019. ACT 2017 has included the national average.*

In [12]:
sat_2017.head() 

State Participation  Evidence-Based Reading and Writing  Math  Total
0     Alabama            5%                                 593   572   1165
1      Alaska           38%                                 547   533   1080
2     Arizona           30%                                 563   553   1116
3    Arkansas            3%                                 614   594   1208
4  California           53%                                 531   524   1055

In [13]:
sat_2018.head()

State Participation  Evidence-Based Reading and Writing  Math  Total
0     Alabama            6%                                 595   571   1166
1      Alaska           43%                                 562   544   1106
2     Arizona           29%                                 577   572   1149
3    Arkansas            5%                                 592   576   1169
4  California           60%                                 540   536   1076

In [14]:
sat_2019.head() 

State Participation Rate  EBRW  Math  Total
0     Alabama                 7%   583   560   1143
1      Alaska                41%   556   541   1097
2     Arizona                31%   569   565   1134
3    Arkansas                 6%   582   559   1141
4  California                63%   534   531   1065

*SAT 2019 has a different column header - EBRW. Renaming of column headers can be carried out.*

*Next, we will use .info() function to check the data types and for missing values:*

In [15]:
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   State          52 non-null     object 
 1   Participation  52 non-null     object 
 2   English        52 non-null     float64
 3   Math           52 non-null     float64
 4   Reading        52 non-null     float64
 5   Science        52 non-null     float64
 6   Composite      52 non-null     object 
dtypes: float64(4), object(3)
memory usage: 3.0+ KB


In [16]:
act_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   State          52 non-null     object 
 1   Participation  52 non-null     object 
 2   Composite      52 non-null     float64
dtypes: float64(1), object(2)
memory usage: 1.3+ KB


In [17]:
act_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   State          52 non-null     object 
 1   Participation  52 non-null     object 
 2   Composite      52 non-null     float64
dtypes: float64(1), object(2)
memory usage: 1.3+ KB


*The composite score in ACT 2017 is saved as an object data-type, similarly to the participation rates from 2017 to 2019. These values can be converted to float data-type.*

*There are 52 counts in each column. However, there are 50 US States + DC (total 51). This can be reviewed later on.*

In [18]:
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   State                               51 non-null     object
 1   Participation                       51 non-null     object
 2   Evidence-Based Reading and Writing  51 non-null     int64 
 3   Math                                51 non-null     int64 
 4   Total                               51 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [19]:
sat_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   State                               51 non-null     object
 1   Participation                       51 non-null     object
 2   Evidence-Based Reading and Writing  51 non-null     int64 
 3   Math                                51 non-null     int64 
 4   Total                               51 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [20]:
sat_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53 entries, 0 to 52
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   State               53 non-null     object
 1   Participation Rate  53 non-null     object
 2   EBRW                53 non-null     int64 
 3   Math                53 non-null     int64 
 4   Total               53 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.2+ KB


*Similarly, the participation rate can be converted to float type. The number of rows (or states) should be checked as well.*

### Observations
There are data cleaning required on the imported datasets: <br>
1. Remove unnecessary columns (e.g. subject results in act_2017)
2. Rename all column headers appropriately (using same headers, changing to lowercase)
3. Convert all data types appropriately (e.g. string to float)
4. Remove unnecessary rows, in order to compare across the same number of states
5. Merge the datasets into 1 dataframe

#### Before cleaning: act_2017 

In [21]:
act_2017.head()

State Participation  English  Math  Reading  Science Composite
0  National           60%     20.3  20.7     21.4     21.0      21.0
1   Alabama          100%     18.9  18.4     19.7     19.4      19.2
2    Alaska           65%     18.7  19.8     20.4     19.9      19.8
3   Arizona           62%     18.6  19.8     20.1     19.8      19.7
4  Arkansas          100%     18.9  19.0     19.7     19.5      19.4

In [22]:
# Removing 'English', 'Math', 'Reading', 'Science' columns:

act_2017.drop(['English', 'Math', 'Reading', 'Science'], axis = 1, inplace = True)

In [23]:
# Removing 'National' row:

act_2017.drop(act_2017.loc[act_2017['State'] == 'National'].index, axis = 0, inplace = True)

In [24]:
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 1 to 51
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   State          51 non-null     object
 1   Participation  51 non-null     object
 2   Composite      51 non-null     object
dtypes: object(3)
memory usage: 1.3+ KB


In [25]:
# # Converting numeric strings to floats:

# act_2017['Composite'] = act_2017['Composite'].astype(float)

We encounted an error when converting `Composite` to float type: <br>
`ValueError: could not convert string to float: '20.2x'`

In [26]:
# Checking for row with error
act_2017.loc[act_2017['Composite'] == '20.2x']

State Participation Composite
51  Wyoming          100%     20.2x

In [27]:
# Remove additonal 'x' found inside Composite score:

act_2017['Composite'] = act_2017.Composite.str.replace('x', '')

In [28]:
# Converting numeric strings to floats:

act_2017['Composite'] = act_2017['Composite'].astype(float)

In [29]:
act_2017['Participation'] = act_2017['Participation'].map(convert_to_float)

In [30]:
# Renaming column headers to lowercase:

act_2017.rename(columns = {
    'State': 'state','Participation': 'part_act_2017', 'Composite': 'comp_act_2017'
}, inplace = True)

In [31]:
# Reset index:

act_2017.reset_index(drop = True, inplace = True)

#### After cleaning: act_2017

In [32]:
act_2017.head()

state  part_act_2017  comp_act_2017
0     Alabama           1.00           19.2
1      Alaska           0.65           19.8
2     Arizona           0.62           19.7
3    Arkansas           1.00           19.4
4  California           0.31           22.8

#### Before cleaning: act_2018

In [33]:
act_2018.head()

State Participation  Composite
0     Alabama          100%       19.1
1      Alaska           33%       20.8
2     Arizona           66%       19.2
3    Arkansas          100%       19.4
4  California           27%       22.7

In [34]:
act_2018.duplicated().sum()

1

In [35]:
act_2018.value_counts()

State                 Participation  Composite
Maine                 7%             24.0         2
Alabama               100%           19.1         1
Montana               100%           20.0         1
Nevada                100%           17.7         1
New Hampshire         16%            25.1         1
New Jersey            31%            23.7         1
New Mexico            67%            19.4         1
New York              27%            24.5         1
North Carolina        100%           19.1         1
North Dakota          98%            20.3         1
Ohio                  100%           20.3         1
Oklahoma              100%           19.3         1
Oregon                42%            21.3         1
Pennsylvania          20%            23.5         1
Rhode Island          15%            24.2         1
South Carolina        100%           18.3         1
South Dakota          77%            21.9         1
Tennessee             100%           19.6         1
Texas            

In [36]:
# Identified duplicate row:

act_2018.loc[(act_2018['State'] == 'Maine')]

State Participation  Composite
19  Maine            7%       24.0
20  Maine            7%       24.0

In [37]:
# Removing duplicate row:

act_2018 = act_2018.drop_duplicates()

In [38]:
# Convert numeric string to float and rename column headers:
    
act_2018['Participation'] = act_2018['Participation'].map(convert_to_float)

In [39]:
act_2018.rename(columns = {
    'State': 'state','Participation': 'part_act_2018', 'Composite': 'comp_act_2018'}
, inplace = True)

In [40]:
# Rename State to 'District of Columbia:

act_2018.loc[8, 'state'] = 'District of Columbia'

In [41]:
# Reset index:

act_2018.reset_index(drop = True, inplace = True)

#### After cleaning: act_2018

In [42]:
act_2018.head()

state  part_act_2018  comp_act_2018
0     Alabama           1.00           19.1
1      Alaska           0.33           20.8
2     Arizona           0.66           19.2
3    Arkansas           1.00           19.4
4  California           0.27           22.7

#### Before cleaning: act_2019

In [43]:
act_2019.head()

State Participation  Composite
0     Alabama          100%       18.9
1      Alaska           38%       20.1
2     Arizona           73%       19.0
3    Arkansas          100%       19.3
4  California           23%       22.6

In [44]:
# Remove 'National' row:

act_2019.drop(act_2019.loc[act_2019['State'] == 'National'].index, axis = 0, inplace = True)

In [45]:
# Convert numeric string to float and rename columns headers:

act_2019['Participation'] = act_2019['Participation'].map(convert_to_float)

In [46]:
act_2019.rename(columns = {
    'State': 'state','Participation': 'part_act_2019', 'Composite': 'comp_act_2019'}
, inplace = True)

In [47]:
#Reset index:

act_2019.reset_index(drop = True, inplace = True)

#### After cleaning: act_2019

In [48]:
act_2019.head()

state  part_act_2019  comp_act_2019
0     Alabama           1.00           18.9
1      Alaska           0.38           20.1
2     Arizona           0.73           19.0
3    Arkansas           1.00           19.3
4  California           0.23           22.6

#### Before cleaning: sat_2017

In [49]:
sat_2017.head()

State Participation  Evidence-Based Reading and Writing  Math  Total
0     Alabama            5%                                 593   572   1165
1      Alaska           38%                                 547   533   1080
2     Arizona           30%                                 563   553   1116
3    Arkansas            3%                                 614   594   1208
4  California           53%                                 531   524   1055

In [50]:
sat_2017.loc[(sat_2017['State'] == 'Maryland')]

State Participation  Evidence-Based Reading and Writing  Math  Total
20  Maryland           69%                                 536    52   1060

*We find that the mean Math score for Maryland is provided as 52. However, the range for Math score is between 200 and 800. It is likely that there is a data entry error. <br>
A score of 524 would make sense as it adds up to a total score of 1060. It also falls between the distribution of Math scores across US states in 2017*

In [51]:
sat_2018['Math'].describe()

count     51.000000
mean     556.235294
std       47.772623
min      480.000000
25%      522.500000
50%      544.000000
75%      593.500000
max      655.000000
Name: Math, dtype: float64

*We will update the mean Math score for Maryland to 524:*

In [52]:
sat_2017.loc[20, ['State', 'Math']] = ['Maryland', 524]

In [53]:
sat_2017.loc[(sat_2017['State'] == 'Maryland')]

State Participation  Evidence-Based Reading and Writing  Math  Total
20  Maryland           69%                                 536   524   1060

In [54]:
# Convert numeric string to float and rename column headers:

sat_2017['Participation'] = sat_2017['Participation'].map(convert_to_float)

In [55]:
sat_2017.rename(columns = {
    'State': 'state',
    'Participation': 'part_sat_2017', 
    'Evidence-Based Reading and Writing': 'read_write_2017', 
    'Math': 'math_2017', 
    'Total': 'total_2017'}, inplace = True)

#### After cleaning: sat_2017

In [56]:
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   state            51 non-null     object 
 1   part_sat_2017    51 non-null     float64
 2   read_write_2017  51 non-null     int64  
 3   math_2017        51 non-null     int64  
 4   total_2017       51 non-null     int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB


#### Before cleaning: sat_2018

In [57]:
sat_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   State                               51 non-null     object
 1   Participation                       51 non-null     object
 2   Evidence-Based Reading and Writing  51 non-null     int64 
 3   Math                                51 non-null     int64 
 4   Total                               51 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [58]:
# Convert numeric string to float and rename column headers:

sat_2018['Participation'] = sat_2018['Participation'].map(convert_to_float)

In [59]:
sat_2018.rename(columns = {
    'State': 'state',
    'Participation': 'part_sat_2018', 
    'Evidence-Based Reading and Writing': 'read_write_2018', 
    'Math': 'math_2018', 
    'Total': 'total_2018'}, inplace = True)

#### After cleaning: sat_2018

In [60]:
sat_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   state            51 non-null     object 
 1   part_sat_2018    51 non-null     float64
 2   read_write_2018  51 non-null     int64  
 3   math_2018        51 non-null     int64  
 4   total_2018       51 non-null     int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB


#### Before cleaning: sat_2019

In [61]:
sat_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53 entries, 0 to 52
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   State               53 non-null     object
 1   Participation Rate  53 non-null     object
 2   EBRW                53 non-null     int64 
 3   Math                53 non-null     int64 
 4   Total               53 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.2+ KB


In [62]:
# Checking for additional rows
set(sat_2019['State'])^(set(sat_2018['state']))

{'Puerto Rico', 'Virgin Islands'}

In [63]:
set(sat_2019['State'])^(set(sat_2017['state']))

{'Puerto Rico', 'Virgin Islands'}

*We will remove 'Puerto Rico' and 'Virgin Islands' from the dataset as the participation rate is not available. <br>
These 2 states are not present in 2017 and 2018 datasets as well, and will not be included in our analysis.*

In [64]:
sat_2019.drop(sat_2019.loc[sat_2019['State'] == 'Puerto Rico'].index, axis = 0, inplace = True)

In [65]:
sat_2019.drop(sat_2019.loc[sat_2019['State'] == 'Virgin Islands'].index, axis = 0, inplace = True)

In [66]:
# Convert numeric string to float and rename column headers:

sat_2019['Participation Rate'] = sat_2019['Participation Rate'].map(convert_to_float)

In [67]:
sat_2019.rename(columns = {
    'State': 'state',
    'Participation Rate': 'part_sat_2019', 
    'EBRW': 'read_write_2019', 
    'Math': 'math_2019', 
    'Total': 'total_2019'}, inplace = True)

In [68]:
# Reset index:

sat_2019.reset_index(drop = True, inplace = True)

#### After cleaning: sat_2019

In [69]:
sat_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   state            51 non-null     object 
 1   part_sat_2019    51 non-null     float64
 2   read_write_2019  51 non-null     int64  
 3   math_2019        51 non-null     int64  
 4   total_2019       51 non-null     int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB


*We will now merge all datasets to 1 dataframe.*

In [70]:
merged_df = act_2017.merge(
    act_2018, on=['state'],how='outer').merge(
    act_2019,on=['state'],how='outer').merge(
    sat_2017,on=['state'],how='outer').merge(
    sat_2018,on=['state'],how='outer').merge(
    sat_2019,on=['state'],how='outer')

In [71]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 51 entries, 0 to 50
Data columns (total 19 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   state            51 non-null     object 
 1   part_act_2017    51 non-null     float64
 2   comp_act_2017    51 non-null     float64
 3   part_act_2018    51 non-null     float64
 4   comp_act_2018    51 non-null     float64
 5   part_act_2019    51 non-null     float64
 6   comp_act_2019    51 non-null     float64
 7   part_sat_2017    51 non-null     float64
 8   read_write_2017  51 non-null     int64  
 9   math_2017        51 non-null     int64  
 10  total_2017       51 non-null     int64  
 11  part_sat_2018    51 non-null     float64
 12  read_write_2018  51 non-null     int64  
 13  math_2018        51 non-null     int64  
 14  total_2018       51 non-null     int64  
 15  part_sat_2019    51 non-null     float64
 16  read_write_2019  51 non-null     int64  
 17  math_2019        5

In [72]:
# Save merged dataframe to new csv
merged_df.to_csv('../data/merged_df.csv', index=False)